src/ore/src/ssh_key.rs

```diff
@@ -36,37 +39,139 @@ impl SshKeypair {
     pub fn new() -> Result<SshKeypair, anyhow::Error> {
         let openssl_key = PKey::<Private>::generate_ed25519()?;
 
-        let keypair = KeypairData::Ed25519(Ed25519Keypair {
+        let keypair_data = KeypairData::Ed25519(Ed25519Keypair {
             public: Ed25519PublicKey::try_from(openssl_key.raw_public_key()?.as_slice())?,
             private: Ed25519PrivateKey::try_from(openssl_key.raw_private_key()?.as_slice())?,
         });
 
-        let private_key_wrapper = PrivateKey::new(keypair, "materialize")?;
-        let openssh_private_key = &*private_key_wrapper.to_openssh(LineEnding::LF)?;
-        let openssh_public_key = private_key_wrapper.public_key().to_openssh()?;
+        let keypair = PrivateKey::new(keypair_data, "materialize")?;
+
+        Ok(SshKeypair { keypair })
+    }
+
+    /// Deserialize a keypair from a private key
+    ///
+    /// In the OpenSSH format, the public key is embedded inside the private key
+    fn from_private_key(private_key: &[u8]) -> Result<SshKeypair, anyhow::Error> {
+        let private_key = PrivateKey::from_openssh(private_key)?;
 
         Ok(SshKeypair {
-            public_key: openssh_public_key.as_bytes().to_vec(),
-            private_key: openssh_private_key.as_bytes().to_vec(),
+            keypair: private_key,
         })
     }
 
     /// Return the public key encoded in the OpenSSH format
-    pub fn ssh_public_key(&self) -> &[u8] {
-        &self.public_key
+    pub fn ssh_public_key(&self) -> String {
+        self.keypair.public_key().to_string()
     }
 
     /// Return the private key encoded in the OpenSSH format
-    pub fn ssh_private_key(&self) -> &[u8] {
-        &self.private_key
+    pub fn ssh_private_key(&self) -> Zeroizing<String> {
+        self.keypair.to_openssh(LineEnding::LF).expect("TODO")
+        // &self.private_key
+    }
+}
+
+impl PartialEq for SshKeypair {
+    fn eq(&self, other: &Self) -> bool {
+        self.keypair.fingerprint(Default::default())
```

Comment:
what does this `fingerprint` do?

Reply:
In broad strokes, it's the SHA256 hash of the key. The fingerprint is what is actually used to identify keys when opening SSH sessions, and it felt like a good enough basis for `PartialEq` (which is only used in tests), as `PrivateKey` doesn't implement it.